File: sorting/실패율.py
# 실패율

from copy import deepcopy


def solution(N, stages):
    data = [0] * (N + 2)
    for stage in stages:
        data[stage] += 1

    accum = deepcopy(data)
    for i in range(len(accum) - 1, 0, -1):
        accum[i - 1] += accum[i]

    data = data[1:-1]
    accum = accum[1:-1]

    result = []
    for i in range(len(data)):
        result.append((data[i] / accum[i] if accum[i] != 0 else 0, i + 1))

    result.sort(key=lambda x: (-x[0], x[1]))
    return [x[1] for x in result]

File: programmers/level1/실패율.js
// https://school.programmers.co.kr/learn/courses/30/lessons/42889

function solution(N, stages) {
  const result = [];
  for (let i = 1; i <= N; i++) {
    const reach = stages.filter((stage) => stage >= i).length;
    const curr = stages.filter((stage) => stage === i).length;
    const calc = reach === 0 ? 0 : curr / reach;
    result.push([i, calc]); // [스테이지 번호, 실패율]
  }

  return result
    .sort((a, b) => {
      if (a[1] !== b[1]) return b[1] - a[1]; // 실패율 내림차순
      return a[0] - b[0]; // 스테이지 번호 오름차순
    })
    .map((elem) => elem[0]);
}
